Use cookie for banners state

obulat opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->
Fixes #976 by @sarayourfriend

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->
This PR adds a UI cookie to save the IDs of banners that have been dismissed by the user. 

## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->
Try opening `http://localhost:8443/ar/search?q=cat&referrer=creativecommons.org`. You should see 2 banners at the top of the window. If you dismiss one of them and reload, it should not re-appear. If you clear the cookies for localhost and reload, it should re-appear.

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

zackkrida reviewed (changes requested):
Not sure if this is ready, but dismissing the banner does not work for me. Additionally I see the following warning about the SameSite property on the cookie:

![CleanShot 2022-11-15 at 11 17 26](https://user-images.githubusercontent.com/6351754/201970997-64c20d7c-8b54-49d8-a6f1-0df8e4ffd8a6.png)

obulat commented:
> Not sure if this is ready, but dismissing the banner does not work for me.

Sorry, I should have drafted this PR until I updated it with the latest changes.

> Additionally, I see the following warning about the SameSite property on the cookie:

Thanks for catching, it was caused by the `middleware` not using cookie options. Fixed!

sarayourfriend commented on src/components/VNotificationBanner.vue:
```diff
@@ -59,17 +58,25 @@ export default defineComponent({
+    /**
+     * We only hide the enabled banner if there is a cookie.
+     */
+    const shouldShow = ref(!uiStore.isBannerDismissed(props.id))
```
This is definitely just a nit about how to manage refs and store-derived data: Would it make sense to call `isBannerDismissed` inside of a computed callback? That way the `dismissBanner` function doesn't have to futz with updating the ref:

```ts
const shouldShow = computed(() => !uiStore.isBannerDismissed(props.id))
```

sarayourfriend commented on src/stores/ui.ts:
```diff
@@ -1,9 +1,12 @@
+
+import type { Locale } from 'vue-i18n'
+
+export type BannerId = Locale | 'cc-referral'
```
The translation banner ids are actually `` `translation-${Locale}` `` not just the locale string.

sarayourfriend commented on src/stores/ui.ts:
```diff
@@ -139,5 +141,16 @@ export const useUiStore = defineStore('ui', {
     toggleFilters() {
       this.setFiltersState(!this.isFilterVisible)
     },
+    dismissBanner(bannerId: string) {
+      this.dismissedBanners.push(bannerId)
+      this.$nuxt.$cookies.set(
+        'uiDismissedBanners',
+        this.dismissedBanners,
+        cookieOptions
+      )
+    },
+    isBannerDismissed(bannerId: string): boolean {
+      return this.dismissedBanners.includes(bannerId)
+    },
```
I suppose these could use the `BannerId` type as well?

```suggestion
    dismissBanner(bannerId: BannerId) {
      this.dismissedBanners.push(bannerId)
      this.$nuxt.$cookies.set(
        'uiDismissedBanners',
        this.dismissedBanners,
        cookieOptions
      )
    },
    isBannerDismissed(bannerId: BannerId): boolean {
      return this.dismissedBanners.includes(bannerId)
    },
```

sarayourfriend reviewed (changes requested):
Works well for me locally! From my perspective the only thing missing are updates to the e2e test that exercise the cookie and unit tests for the new store methods.

The rest of my feedback is just nit-picks and can be safely ignored, in my opinion.

obulat commented:
> Works well for me locally! From my perspective, the only thing missing are updates to the e2e test that exercise the cookie and unit tests for the new store methods.

While adding some code review suggestions, I realized that just like the layout properties, it is best to keep everything inside the `ui` store instead of splitting the state between composables and the store. So, I've re-written the logic and extracted the banners out of the layouts into a separate `VBanners` component. Sorry for doubling the review work.

sarayourfriend reviewed (approved):
LGTM! :rocket:

obulat commented on src/components/VBanner/VBanners.vue:
```diff
@@ -0,0 +1,48 @@
+export default defineComponent({
+  name: 'VBanners',
+  components: {
+    VMigrationNotice: () => import('~/components/VBanner/VMigrationNotice.vue'),
+    VTranslationStatusBanner: () =>
+      import('~/components/VBanner/VTranslationStatusBanner.vue'),
```
It allows us to lazily load components. It's probably not as important for small components. We can also add the webpack comment to allow prefetching, although for the banners it doesn't make sense: the banner will only be displayed if the user goes to a language with fewer translated strings, but we would prefetch the code for everyone?

More info about it here:
https://vueschool.io/articles/vuejs-tutorials/lazy-loading-individual-vue-components-and-prefetching/

It would be nice to set up a way of measuring performance to see if things like this are really necessary or not.

zackkrida commented on src/components/VBanner/VBanners.vue:
```diff
@@ -0,0 +1,48 @@
+      @close="dismissBanner(translationBannerId)"
+    />
+  </div>
+</template>
+<script lang="ts">
```
```suggestion
</template>

<script lang="ts">
```

zackkrida commented on src/utils/translation-banner.ts:
```diff
@@ -0,0 +1,12 @@
+export const needsTranslationBanner = (locale: LocaleObject) => {
+  if (!locale || locale.code === 'en') return false
+
+  return (locale.translated ?? 100) <= 90
```
I might prefer if 90 was captured in a variable called like `MINIMUM_TRANSLATION_PERCENTAGE` or if there was a comment explaining what it represents. I don't love having an arbitrary number like that without context.

zackkrida commented on src/stores/ui.ts:
```diff
@@ -139,5 +178,22 @@ export const useUiStore = defineStore('ui', {
+     * @param bannerId - the id of the banner to dismiss.
+     */
+    dismissBanner(bannerId: BannerId) {
+      if (!this.dismissedBanners.includes(bannerId)) {
+        this.dismissedBanners.push(bannerId)
+        this.$nuxt.$cookies.set(
+          'uiDismissedBanners',
+          this.dismissedBanners,
+          cookieOptions
+        )
+      }
```
```suggestion
      if (!this.dismissedBanners.includes(bannerId)) {
        return
      }
      
      this.dismissedBanners.push(bannerId)
      this.$nuxt.$cookies.set(
        'uiDismissedBanners',
        this.dismissedBanners,
        cookieOptions
      )
```

Personally prefer early returns but it doesn't matter.

zackkrida commented on src/stores/ui.ts:
```diff
@@ -28,13 +34,10 @@ export interface UiState {
-  sameSite: 'strict',
-  maxAge: 60 * 60 * 24 * 60, // 60 days
-  secure: isProd,
+  /**
+   * array of banner ids that were dismissed
+   */
```
```suggestion
  /**
   * array of banner ids that were dismissed
   */
```

Not a necessary comment, imo.
